File: sdk/sms.js
/**
 * SMS module - SMS Sender and Receiver classes
 * Equivalent to PHP SMSSender and SMSReceiver classes in sdk_file.php
 */

const Core = require('./core');

// ─── Custom Exception ───────────────────────────────────────────────
class SMSServiceException extends Error {
  constructor(message, statusCode) {
    super(message);
    this.name = 'SMSServiceException';
    this.statusCode = statusCode;
    this.statusDetail = message;
  }

  getErrorCode() {
    return this.statusCode;
  }

  getErrorMessage() {
    return this.statusDetail;
  }
}

// ─── SMS Receiver ───────────────────────────────────────────────────
class SMSReceiver {
  /**
   * Parse incoming SMS notification JSON
   * @param {object} jsonRequest - Parsed JSON body from incoming request
   */
  constructor(jsonRequest) {
    if (!jsonRequest || (!jsonRequest.sourceAddress && !jsonRequest.message)) {
      this.response = { statusCode: 'E1312', statusDetail: 'Request is Invalid.' };
    } else {
      this.thejson = jsonRequest;
      this.version = jsonRequest.version || null;
      this.applicationId = jsonRequest.applicationId || null;
      this.sourceAddress = jsonRequest.sourceAddress || jsonRequest.address || null;
      this.message = jsonRequest.message || null;
      this.requestId = jsonRequest.requestId || null;
      this.encoding = jsonRequest.encoding || null;
      this.response = { statusCode: 'S1000', statusDetail: 'Process completed successfully.' };
    }
  }

  getVersion() { return this.version; }
  getEncoding() { return this.encoding; }
  getApplicationId() { return this.applicationId; }
  getAddress() { return this.sourceAddress; }
  getMessage() { return this.message; }
  getRequestId() { return this.requestId; }
  getJson() { return this.thejson; }
  getResponse() { return this.response; }
}

// ─── SMS Sender ─────────────────────────────────────────────────────
class SMSSender extends Core {
  /**
   * @param {string} serverURL - BDApps SMS send URL
   * @param {string} applicationId - Application ID
   * @param {string} password - Application password
   */
  constructor(serverURL, applicationId, password) {
    super();

    if (!serverURL || !applicationId || !password) {
      throw new SMSServiceException('Request Invalid.', 'E1312');
    }

    this.applicationId = applicationId;
    this.password = password;
    this.serverURL = serverURL;

    // Optional fields
    this.sourceAddress = null;
    this.chargingAmount = null;
    this.encoding = null;
    this.version = null;
    this.deliveryStatusRequest = null;
    this.binaryHeader = null;
  }

  /**
   * Broadcast a message to all subscribed users
   * @param {string} message - Message to broadcast
   * @returns {Promise<boolean>}
   */
  async broadcast(message) {
    return this.sms(message, ['tel:all']);
  }

  /**
   * Send SMS to one or more addresses
   * @param {string} message - Message to send
   * @param {string|string[]} addresses - Destination address(es)
   * @returns {Promise<boolean>}
   */
  async sms(message, addresses) {
    if (!addresses || (Array.isArray(addresses) && addresses.length === 0)) {
      throw new SMSServiceException('Format of the address is invalid.', 'E1325');
    }

    const addressList = typeof addresses === 'string' ? [addresses] : addresses;
    const jsonStream = this._resolveJsonStream(message, addressList);

    if (jsonStream) {
      const response = await this.sendRequest(JSON.parse(jsonStream), this.serverURL);
      return this._handleResponse(response);
    }

    return false;
  }

  /** @private */
  _handleResponse(jsonResponse) {
    if (!jsonResponse) {
      throw new SMSServiceException('Invalid server URL', '500');
    }

    const { statusCode, statusDetail } = jsonResponse;

    if (statusCode === 'S1000') {
      return true;
    }

    throw new SMSServiceException(statusDetail || 'Unknown error', statusCode || 'UNKNOWN');
  }

  /** @private */
  _resolveJsonStream(message, addresses) {
    const messageDetails = {
      message,
      destinationAddresses: addresses,
    };

    if (this.sourceAddress) messageDetails.sourceAddress = this.sourceAddress;
    if (this.deliveryStatusRequest) messageDetails.deliveryStatusRequest = this.deliveryStatusRequest;
    if (this.binaryHeader) messageDetails.binaryHeader = this.binaryHeader;
    if (this.version) messageDetails.version = this.version;
    if (this.encoding) messageDetails.encoding = this.encoding;

    const payload = {
      applicationId: this.applicationId,
      password: this.password,
      ...messageDetails,
    };

    return JSON.stringify(payload);
  }

  // Setters
  setSourceAddress(sourceAddress) { this.sourceAddress = sourceAddress; }
  setChargingAmount(chargingAmount) { this.chargingAmount = chargingAmount; }
  setEncoding(encoding) { this.encoding = encoding; }
  setVersion(version) { this.version = version; }
  setBinaryHeader(binaryHeader) { this.binaryHeader = binaryHeader; }
  setDeliveryStatusRequest(deliveryStatusRequest) { this.deliveryStatusRequest = deliveryStatusRequest; }
}

module.exports = { SMSSender, SMSReceiver, SMSServiceException };

File: python/sdk/sms.py
"""
SMS module - SMS Sender and Receiver classes.
Equivalent to PHP SMSSender and SMSReceiver classes in sdk_file.php.
"""

import json
from typing import Optional, Union

from .core import Core
from .exceptions import BDAppsException


class SMSServiceException(BDAppsException):
    """Exception for SMS service errors."""
    pass


class SMSReceiver:
    """
    Parse incoming SMS notification JSON.
    Equivalent to PHP SMSReceiver class.
    """

    def __init__(self, json_request: dict):
        if not json_request or (
            "sourceAddress" not in json_request and "message" not in json_request
        ):
            self.response = {
                "statusCode": "E1312",
                "statusDetail": "Request is Invalid.",
            }
            self.thejson = None
            self.version = None
            self.application_id = None
            self.source_address = None
            self.message = None
            self.request_id = None
            self.encoding = None
        else:
            self.thejson = json_request
            self.version = json_request.get("version")
            self.application_id = json_request.get("applicationId")
            self.source_address = json_request.get(
                "sourceAddress", json_request.get("address")
            )
            self.message = json_request.get("message")
            self.request_id = json_request.get("requestId")
            self.encoding = json_request.get("encoding")
            self.response = {
                "statusCode": "S1000",
                "statusDetail": "Process completed successfully.",
            }

    def get_version(self) -> Optional[str]:
        return self.version

    def get_encoding(self) -> Optional[str]:
        return self.encoding

    def get_application_id(self) -> Optional[str]:
        return self.application_id

    def get_address(self) -> Optional[str]:
        return self.source_address

    def get_message(self) -> Optional[str]:
        return self.message

    def get_request_id(self) -> Optional[str]:
        return self.request_id

    def get_json(self) -> Optional[dict]:
        return self.thejson

    def get_response(self) -> dict:
        return self.response


class SMSSender(Core):
    """
    Send SMS to one or more addresses.
    Equivalent to PHP SMSSender class.
    """

    def __init__(self, server_url: str, application_id: str, password: str):
        super().__init__()

        if not server_url or not application_id or not password:
            raise SMSServiceException("Request Invalid.", "E1312")

        self.application_id = application_id
        self.password = password
        self.server_url = server_url

        # Optional fields
        self.source_address: Optional[str] = None
        self.charging_amount: Optional[str] = None
        self.encoding: Optional[str] = None
        self.version: Optional[str] = None
        self.delivery_status_request: Optional[str] = None
        self.binary_header: Optional[str] = None

    async def broadcast(self, message: str) -> bool:
        """Broadcast a message to all subscribed users."""
        return await self.sms(message, ["tel:all"])

    async def sms(self, message: str, addresses: Union[str, list]) -> bool:
        """
        Send SMS to one or more addresses.

        Args:
            message: Message to send.
            addresses: Single address string or list of addresses.

        Returns:
            True if successful.
        """
        if not addresses:
            raise SMSServiceException("Format of the address is invalid.", "E1325")

        address_list = [addresses] if isinstance(addresses, str) else addresses
        json_stream = self._resolve_json_stream(message, address_list)

        if json_stream:
            response = await self.send_request(json.loads(json_stream), self.server_url)
            return self._handle_response(response)

        return False

    def _handle_response(self, json_response: dict) -> bool:
        if not json_response:
            raise SMSServiceException("Invalid server URL", "500")

        status_code = json_response.get("statusCode", "")
        status_detail = json_response.get("statusDetail", "Unknown error")

        if status_code == "S1000":
            return True

        raise SMSServiceException(status_detail, status_code)

    def _resolve_json_stream(self, message: str, addresses: list) -> str:
        message_details = {
            "message": message,
            "destinationAddresses": addresses,
        }

        if self.source_address:
            message_details["sourceAddress"] = self.source_address
        if self.delivery_status_request:
            message_details["deliveryStatusRequest"] = self.delivery_status_request
        if self.binary_header:
            message_details["binaryHeader"] = self.binary_header
        if self.version:
            message_details["version"] = self.version
        if self.encoding:
            message_details["encoding"] = self.encoding

        payload = {
            "applicationId": self.application_id,
            "password": self.password,
            **message_details,
        }

        return json.dumps(payload)

    # Setters
    def set_source_address(self, source_address: str):
        self.source_address = source_address

    def set_charging_amount(self, charging_amount: str):
        self.charging_amount = charging_amount

    def set_encoding(self, encoding: str):
        self.encoding = encoding

    def set_version(self, version: str):
        self.version = version

    def set_binary_header(self, binary_header: str):
        self.binary_header = binary_header

    def set_delivery_status_request(self, delivery_status_request: str):
        self.delivery_status_request = delivery_status_request
